### ISSUE TITLE
Native Rust AD: Hessian-vector product demonstration

### ISSUE THREAD

**Issue by shinaoka** [shinaoka/ndtensors-rs#56]
## Parent Issue
Part of #52 (AD Roadmap), specifically Phase 4: Hessian-vector products.

## Dependencies
- #54 - Reverse-mode PoC (`TrackedTensor` + `backward()`)
- #55 - Forward-mode PoC (`DualTensor` + JVP)

## Goal

Demonstrate Hessian-vector product computation using forward-on-reverse composition.

**Mathematical background**:
```
Given: f: R^n → R (scalar function)
Want:  H @ v  where H = ∇²f (Hessian matrix)

Method (forward-on-reverse):
  1. g(x) = ∇f(x)           # gradient function (reverse-mode)
  2. H @ v = Jg(x) @ v      # JVP of gradient (forward-mode)
  
Cost: O(1) Hessian-vector products (no explicit H construction)
```

### API

```rust
// crates/ndtensors/src/autodiff/hessian.rs

/// Compute Hessian-vector product: H @ v where H = ∇²f(x)
/// 
/// Uses forward-on-reverse: JVP of the gradient function
/// 
/// # Arguments
/// * `f` - Scalar function f: Tensor → Scalar
/// * `x` - Point at which to evaluate
/// * `v` - Vector to multiply with Hessian
/// 
/// # Returns
/// * `H @ v` - Hessian-vector product
pub fn hvp<T, F>(
    f: F,
    x: &DenseTensor<T>,
    v: &DenseTensor<T>,
) -> Result<DenseTensor<T>, TensorError>
where
    T: Scalar,
    F: Fn(&TrackedTensor<T>) -> TrackedTensor<T>;
```

### Implementation Strategy

```rust
pub fn hvp<T, F>(f: F, x: &DenseTensor<T>, v: &DenseTensor<T>) -> Result<DenseTensor<T>, TensorError>
where
    T: Scalar,
    F: Fn(&TrackedTensor<T>) -> TrackedTensor<T>,
{
    // Step 1: Define gradient function g(x) = ∇f(x)
    // This uses reverse-mode AD (#54)
    let grad_at = |x_val: &DenseTensor<T>| -> DenseTensor<T> {
        let x_tracked = TrackedTensor::from_tensor(x_val.clone(), true);
        let y = f(&x_tracked);
        let grads = backward(&y)?;
        grads.get(&x_tracked)?
    };
    
    // Step 2: Compute JVP of gradient function
    // This uses forward-mode AD (#55)
    // 
    // We need: d/dt [∇f(x + t*v)] at t=0
    //        = Jg(x) @ v
    //        = H @ v
    
    let x_dual = DualTensor::new(x.clone(), v.clone());
    
    // Apply gradient function with dual numbers
    // This requires lifting grad_at to work with DualTensor
    let grad_dual = dual_gradient(f, &x_dual)?;
    
    Ok(grad_dual.tangent().unwrap_or_else(|| DenseTensor::zeros(x.shape())))
}
```

### Alternative: Explicit Composition

```rust
/// More explicit version showing the forward-on-reverse pattern
pub fn hvp_explicit<T, F>(
    f: F,
    x: &DenseTensor<T>,
    v: &DenseTensor<T>,
) -> Result<DenseTensor<T>, TensorError>
where
    T: Scalar,
    F: Fn(&DualTrackedTensor<T>) -> DualTrackedTensor<T>,  // works with both AD modes
{
    // Create tensor that tracks both:
    // - Reverse-mode graph (for gradient)
    // - Forward-mode tangent (for Hessian direction)
    let x_dual_tracked = DualTrackedTensor::new(x.clone(), v.clone(), true);
    
    // Forward pass builds graph AND propagates tangents
    let y = f(&x_dual_tracked);
    
    // Backward pass computes gradient AND its tangent (= Hvp)
    let (grad, hvp) = backward_with_tangent(&y)?;
    
    Ok(hvp)
}
```

### Test 1: Quadratic Function

```rust
#[test]
fn test_hvp_quadratic() {
    // f(x) = 0.5 * x^T A x  where A is symmetric
    // ∇f(x) = A x
    // H = A
    // H @ v = A @ v
    
    let a = DenseTensor::from_vec(vec![
        2.0, 1.0,
        1.0, 3.0,
    ], &[2, 2]).unwrap();
    
    let x = DenseTensor::from_vec(vec![1.0, 2.0], &[2]).unwrap();
    let v = DenseTensor::from_vec(vec![1.0, 0.0], &[2]).unwrap();
    
    let f = |x: &TrackedTensor<f64>| {
        // 0.5 * x^T A x
        let ax = tracked_matvec(&a, x);
        tracked_dot(x, &ax) * 0.5
    };
    
    let hvp_result = hvp(f, &x, &v).unwrap();
    
    // Expected: A @ v = [2, 1]
    assert_tensor_close(&hvp_result, &DenseTensor::from_vec(vec![2.0, 1.0], &[2]).unwrap());
}
```

### Test 2: Non-quadratic Function

```rust
#[test]
fn test_hvp_nonquadratic() {
    // f(x, y) = x^2 * y + y^3
    // ∇f = [2xy, x^2 + 3y^2]
    // H = [[2y, 2x], [2x, 6y]]
    
    let x = DenseTensor::from_vec(vec![1.0, 2.0], &[2]).unwrap();  // x=1, y=2
    let v = DenseTensor::from_vec(vec![1.0, 1.0], &[2]).unwrap();
    
    let f = |t: &TrackedTensor<f64>| {
        let x = t.get(&[0]);
        let y = t.get(&[1]);
        x * x * y + y * y * y
    };
    
    let hvp_result = hvp(f, &x, &v).unwrap();
    
    // H @ v at (1, 2):
    // [[4, 2], [2, 12]] @ [1, 1] = [6, 14]
    assert_tensor_close(&hvp_result, &DenseTensor::from_vec(vec![6.0, 14.0], &[2]).unwrap());
}
```

### Test 3: Tensor Contraction

```rust
#[test]
fn test_hvp_contraction() {
    // f(A) = ||A @ B||^2 for fixed B
    // This tests HVP with tensor operations
    
    let b = DenseTensor::from_vec(vec![1.0, 2.0, 3.0], &[3]).unwrap();
    
    let f = |a: &TrackedTensor<f64>| {
        let c = tracked_contract(a, &[1, -1], &b, &[-1]);
        tracked_norm_sqr(&c)
    };
    
    let a = DenseTensor::ones(&[2, 3]);
    let v = DenseTensor::from_vec(vec![1.0, 0.0, 0.0, 0.0, 0.0, 0.0], &[2, 3]).unwrap();
    
    let hvp_result = hvp(f, &a, &v).unwrap();
    // Verify against numerical differentiation
}
```

## Implementation Tasks

- [ ] **Task 1**: Design `hvp` function signature
- [ ] **Task 2**: Implement forward-on-reverse composition
- [ ] **Task 3**: Test with quadratic function (exact Hessian known)
- [ ] **Task 4**: Test with tensor contraction operations
- [ ] **Task 5**: Document usage patterns

## File Structure

```
crates/ndtensors/src/autodiff/
├── mod.rs
├── hessian.rs    // hvp function
└── ...
```

## Applications

Hessian-vector products are essential for:

1. **Newton-CG optimization**: Second-order methods without explicit Hessian
2. **Hessian spectrum analysis**: Lanczos/power iteration on H
3. **Uncertainty quantification**: Laplace approximation
4. **Natural gradient**: Fisher information matrix approximation

## References

- JAX HVP: `jax.jvp` composed with `jax.grad`
- PyTorch: `torch.autograd.functional.hvp`
- "Fast Exact Multiplication by the Hessian" (Pearlmutter, 1994)
